File: elixir/lib/2023/08_haunted_wasteland/aoc202308.ex
defmodule AOC2023.Day08 do
  @moduledoc """
  Advent of Code 2023, day 8: Haunted Wasteland.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    [path, nodes] = puzzle_input |> String.split("\n\n", trim: true)

    {path |> String.to_charlist() |> Enum.map(fn char -> if char == ?L, do: 0, else: 1 end),
     nodes
     |> String.replace("(", "")
     |> String.replace(")", "")
     |> String.split("\n", trim: true)
     |> Enum.map(fn line -> line |> String.split(" = ") end)
     |> Enum.map(fn [from, to] -> {from, to |> String.split(", ") |> List.to_tuple()} end)
     |> Enum.into(%{})}
  end

  @doc """
  Solve part 1.
  """
  def part1(data) do
    {path, nodes} = data
    walk_path(nodes, path)
  end

  @doc """
  Solve part 2.
  """
  def part2(data) do
    {path, nodes} = data
    walk_ghost_path(nodes, path)
  end

  @doc """
  Walk a single path from AAA to ZZZ.

  ## Example:

      iex> nodes = %{"AAA" => {"BBB", "BBB"}, "BBB" => {"AAA", "ZZZ"}, "ZZZ" => {"ZZZ", "ZZZ"}}
      iex> walk_path(nodes, [0, 0, 1])
      6
  """
  def walk_path(nodes, directions),
    do: walk_path(nodes, directions, directions, "AAA", 0)

  def walk_path(_nodes, _directions, _path, "ZZZ", num_steps), do: num_steps

  def walk_path(nodes, directions, [], current, num_steps),
    do: walk_path(nodes, directions, directions, current, num_steps)

  def walk_path(nodes, directions, [turn | path], current, num_steps) do
    walk_path(nodes, directions, path, nodes[current] |> elem(turn), num_steps + 1)
  end

  @doc """
  Walk multiple paths from all xxAs to all xxZs.

  ## Example:

      iex> nodes = %{
      ...>     "AAA" => {"11B", "XXX"},
      ...>     "11B" => {"XXX", "ZZZ"},
      ...>     "ZZZ" => {"11B", "XXX"},
      ...>     "22A" => {"22B", "XXX"},
      ...>     "22B" => {"22C", "22C"},
      ...>     "22C" => {"22Z", "22Z"},
      ...>     "22Z" => {"22B", "22B"},
      ...>     "XXX" => {"XXX", "XXX"},
      ...> }
      iex> walk_ghost_path(nodes, [0, 1])
      6
  """
  def walk_ghost_path(nodes, directions) do
    start_nodes = nodes |> Map.keys() |> Enum.filter(fn node -> String.ends_with?(node, "A") end)

    walk_ghost_path(
      nodes,
      directions,
      directions,
      start_nodes,
      start_nodes |> Enum.with_index() |> Enum.map(fn {_, idx} -> {idx, 0} end) |> Enum.into(%{}),
      0
    )
  end

  def walk_ghost_path(nodes, directions, [], current, steps, num_steps),
    do: walk_ghost_path(nodes, directions, directions, current, steps, num_steps)

  def walk_ghost_path(nodes, directions, [turn | path], current, steps, num_steps) do
    steps =
      Enum.reduce(current |> Enum.with_index(), steps, fn {node, idx}, steps ->
        if String.ends_with?(node, "Z"), do: Map.put(steps, idx, num_steps), else: steps
      end)

    if Enum.all?(Map.values(steps), &(&1 > 0)),
      do: steps |> Map.values() |> Enum.reduce(1, fn step, lcm -> Math.lcm(step, lcm) end),
      else:
        walk_ghost_path(
          nodes,
          directions,
          path,
          current |> Enum.map(fn node -> nodes[node] |> elem(turn) end),
          steps,
          num_steps + 1
        )
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2023/08_haunted_wasteland/aoc202308.py
"""AoC 8, 2023: Haunted Wasteland."""

# Standard library imports
import itertools
import math
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    path, nodes = puzzle_input.split("\n\n")

    return (
        [0 if ch == "L" else 1 for ch in path],
        dict(parse_node(node) for node in nodes.split("\n")),
    )


def parse_node(node):
    """Parse one node of input.

    ## Example:

    >>> parse_node("AAA = (BBB, CCC)")
    ('AAA', ('BBB', 'CCC'))
    """
    start, targets = node.split(" = ")
    return (start, tuple(targets.replace("(", "").replace(")", "").split(", ")))


def part1(data):
    """Solve part 1."""
    path, nodes = data
    return walk_path(nodes, path)


def part2(data):
    """Solve part 2."""
    path, nodes = data
    return walk_ghost_path(nodes, path)


def walk_path(nodes, path):
    """Walk a single path from AAA to ZZZ.

    ## Example:

    >>> nodes = {"AAA": ("BBB", "BBB"), "BBB": ("AAA", "ZZZ"), "ZZZ": ("ZZZ", "ZZZ")}
    >>> walk_path(nodes, [0, 0, 1])
    6
    """
    current = "AAA"
    for num_steps, turn in enumerate(itertools.cycle(path)):
        if current == "ZZZ":
            return num_steps
        current = nodes[current][turn]


def walk_ghost_path(nodes, path):
    """Walk multiple paths from all __As to all __Zs.

    ## Example:

    >>> nodes = {
    ...     "AAA": ("11B", "XXX"),
    ...     "11B": ("XXX", "ZZZ"),
    ...     "ZZZ": ("11B", "XXX"),
    ...     "22A": ("22B", "XXX"),
    ...     "22B": ("22C", "22C"),
    ...     "22C": ("22Z", "22Z"),
    ...     "22Z": ("22B", "22B"),
    ...     "XXX": ("XXX", "XXX"),
    ... }
    >>> walk_ghost_path(nodes, [0, 1])
    6
    """
    current = [node for node in nodes if node.endswith("A")]
    steps = [0 for _ in current]
    for num_steps, turn in enumerate(itertools.cycle(path)):
        for idx, node in enumerate(current):
            if node.endswith("Z") and not steps[idx]:
                steps[idx] = num_steps
                if all(steps):
                    return math.lcm(*steps)
        current = [nodes[node][turn] for node in current]


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
